Create a template for all dimensions

thiagoftsm opened the pull request:
##### Summary
Fixes #2107
Fixes #6594 
Fixes #3249

I  started to implement a specific user request, but during the development it was identified a bug in the Netdata, so I am also bringing the solution for the bug that is related with this PR.

Before this PR the user had to copy their alarms changing dimensions when they wanna set alarms for each dimension, with the code here the user now can use the keyword `foreach` to set the dimensions that the lookup rule will be applied.

This PR followed the description given by @cakrit in the PR #2107 to create alarms from template, but considering that @ilyam8 also gave an explanation for our users in how to set alarms to dimensions without to use the templates, the final solution also contemplates this possibility for the users.

I am bringing also 4 alarms examples and a shell script that use them to demonstrate how the newest feature will work.

##### Component Name
Health and Database
##### Additional Information

ilyam8 commented:
@thiagoftsm

Could you describe how it will work in the OP? What keyword will be added, what is you idea? It would be nice to have some examples.

like

```
# make sure apache is running

template: apache_last_collected_secs
      on: apache.requests
    calc: $now - $last_collected_t FOREACH ...
   units: seconds ago
   every: 10s
    warn: $this > (($status >= $WARNING)  ? ($update_every) : ( 5 * $update_every))
    crit: $this > (($status == $CRITICAL) ? ($update_every) : (60 * $update_every))
   delay: down 5m multiplier 1.5 max 1h
    info: number of seconds since the last successful data collection
      to: webmaster
```

thiagoftsm commented:
> @thiagoftsm
> 
> Could you describe how it will work in the OP? What keyword will be added, what is you idea? It would be nice to have some examples.
> 
> like
> 
> ```
> # make sure apache is running
> 
> template: apache_last_collected_secs
>       on: apache.requests
>     calc: $now - $last_collected_t FOREACH ...
>    units: seconds ago
>    every: 10s
>     warn: $this > (($status >= $WARNING)  ? ($update_every) : ( 5 * $update_every))
>     crit: $this > (($status == $CRITICAL) ? ($update_every) : (60 * $update_every))
>    delay: down 5m multiplier 1.5 max 1h
>     info: number of seconds since the last successful data collection
>       to: webmaster
> ```

Hi @ilyam8 ,

I commit few minutes ago the update in the documentation. The example that I am using to develop this PR is:

 alarm: dev_dim_template
    on: system.cpu
    os: linux
lookup: sum -3s at 0 every 3 percentage of softirq,user foreach softirq,user
 units: %
 every: 1s
  warn: $this > 1
  crit: $this > 4

I am following @cakrit suggestion in his last comment in the original issue, so I am changing the lookup.

thiagoftsm commented:
> @thiagoftsm
> 
> Could you describe how it will work in the OP? What keyword will be added, what is you idea? It would be nice to have some examples.
> 
> like
> 
> ```
> # make sure apache is running
> 
> template: apache_last_collected_secs
>       on: apache.requests
>     calc: $now - $last_collected_t FOREACH ...
>    units: seconds ago
>    every: 10s
>     warn: $this > (($status >= $WARNING)  ? ($update_every) : ( 5 * $update_every))
>     crit: $this > (($status == $CRITICAL) ? ($update_every) : (60 * $update_every))
>    delay: down 5m multiplier 1.5 max 1h
>     info: number of seconds since the last successful data collection
>       to: webmaster
> ```

Hi @ilyam8 ,

I commit few minutes ago the update in the documentation. The example that I am using to develop this PR is:

 alarm: dev_dim_template
    on: system.cpu
    os: linux
lookup: sum -3s at 0 every 3 percentage of softirq,user foreach softirq,user
 units: %
 every: 1s
  warn: $this > 1
  crit: $this > 4

I am following @cakrit suggestion in his last comment in the original issue, so I am changing the lookup.

ilyam8 commented:
> lookup: sum -3s at 0 every 3 percentage of softirq,user foreach softirq,user

How this supposed to work? I mean `of` and `foreach` together. And `sum` with `foreach`.

Could you describe the whole line what does it mean

> lookup: sum -3s at 0 every 3 percentage of softirq,user foreach softirq,user

thiagoftsm commented:
> > lookup: sum -3s at 0 every 3 percentage of softirq,user foreach softirq,user
> 
> How this supposed to work? I mean `of` and `foreach` together. And `sum` with `foreach`.
> 
> Could you describe the whole line what does it mean
> 
> > lookup: sum -3s at 0 every 3 percentage of softirq,user foreach softirq,user

This is a development test, so the first thing is  to avoid break the features that is already present, until now I reached this goal.
The main goal of the foreach is to apply the rules to search the information for each dimension listed. I am thinking about the possible contradiction between "of" and "foreach" and decision to take with this, but for sure we cannot remove the "of" due backward compatibility.

ilyam8 commented:
I think no, because it wont work with dynamic dimensions :thinking:

ilyam8 commented:
ok, i tested the PR.

What i did:
 - installed the PR

```
[ilyam@pc netdata_foreach]$ git status 
On branch health_connection
Your branch is up to date with 'origin/health_connection'.

nothing to commit, working tree clean
```

 - configured following alarm

```
[ilyam@pc health.d]$ sudo cat example.conf 
template: example
      on: random
  lookup: max -10s foreach *random*
    warn: $this > (($status >= $WARNING)  ? (75) : (85))
    crit: $this > (($status == $CRITICAL) ? (85) : (95))
      to: sysadmin
```

___

Dimensions:

```json
{
    "random1": {
        "name": "random1"
    },
    "random2": {
        "name": "random2"
    },
    "random3": {
        "name": "random3"
    }
}
```

I got 1 alarm
```
{
    "example.random.example_*random*": {
        "id": 1565354637,
        "name": "example_*random*",
        "chart": "example.random",
        "family": "random",
        "active": true,
        "lookup_dimensions": "*random*",
    }
}
```

Patterns after `foreach` are simple patterns.

ilyam8 commented:
I changed alarm to (removed simple pattern because it doesn work)

```
[ilyam@pc health.d]$ sudo cat example.conf 
template: example
      on: random
  lookup: max -10s match-ids foreach random1,random2,random3
    warn: $this > (($status >= $WARNING)  ? (75) : (85))
    crit: $this > (($status == $CRITICAL) ? (85) : (95))
      to: sysadmin
```

> lookup: max -10s match-ids foreach random1,random2,random3

I expect to to get 3 alarms with following `lookup dimensions`:
 - "lookup_dimensions": "random1"
 - "lookup_dimensions": "random2"
 - "lookup_dimensions": "random2"

I got

```json
    "alarms": {
        "example.random.example_random1": {
            "id": 1565354639,
            "name": "example_random1",
            "chart": "example.random",
            "lookup_dimensions": "random1",
        },
        "example.random.example_random2": {
            "id": 1565354638,
            "name": "example_random2",
            "chart": "example.random",
            "lookup_dimensions": "random1,random2",
        },
        "example.random.example_random3": {
            "id": 1565354640,
            "name": "example_random3",
            "chart": "example.random",
            "lookup_dimensions": "random1,random2,random3",

        }
    }
```

ilyam8 commented:
i didnt mention it - the value is wrong too

```json
        "example.random.example_random1": {
            "id": 1565354639,
            "name": "example_random1",
            "chart": "example.random",
            "lookup_dimensions": "random1",
        },
```

the value is sum of all dimensions

thiagoftsm commented:
> > Template for All dimensions
> 
> I still dont have clear view how it works.
> 
> Example, lets take `template`:
> 
> ```
> template: 10min_cpu_usage
>       on: system.cpu
>   lookup: average -10m unaligned of user,system,softirq
>     warn: $this > (($status >= $WARNING)  ? (75) : (85))
>     crit: $this > (($status == $CRITICAL) ? (85) : (95))
>       to: sysadmin
> ```
> 
> if i change it to
> 
> ```
> template: 10min_cpu_usage
>       on: system.cpu
>   lookup: average -10m unaligned foreach user,system,softirq
>     warn: $this > (($status >= $WARNING)  ? (75) : (85))
>     crit: $this > (($status == $CRITICAL) ? (85) : (95))
>       to: sysadmin
> ```
> 
> it will be internally copied len(foreach) times and the end result will be
> 
> ```
> template: 10min_cpu_usage
>       on: system.cpu
>   lookup: average -10m unaligned of user
>     warn: $this > (($status >= $WARNING)  ? (75) : (85))
>     crit: $this > (($status == $CRITICAL) ? (85) : (95))
>       to: sysadmin
> 
> template: 10min_cpu_usage
>       on: system.cpu
>   lookup: average -10m unaligned of system
>     warn: $this > (($status >= $WARNING)  ? (75) : (85))
>     crit: $this > (($status == $CRITICAL) ? (85) : (95))
>       to: sysadmin
> 
> template: 10min_cpu_usage
>       on: system.cpu
>   lookup: average -10m unaligned of softirq
>     warn: $this > (($status >= $WARNING)  ? (75) : (85))
>     crit: $this > (($status == $CRITICAL) ? (85) : (95))
>       to: sysadmin
> ```
> 
> Is this how it will work or not?

Yes, it is exactly this way we hope it will work. It will get all the other information inside the  alarm and it will copy for the others.

thiagoftsm commented:
> i didnt mention it - the value is wrong too
> 
> ```json
>         "example.random.example_random1": {
>             "id": 1565354639,
>             "name": "example_random1",
>             "chart": "example.random",
>             "lookup_dimensions": "random1",
>         },
> ```
> 
> > "lookup_dimensions": "random1",
> 
> The value is sum of all dimensions.

I identified the problem, I copied and paste the code without to change the header.

> I changed alarm to (removed simple pattern because it doesn work)
> 
> ```
> [ilyam@pc health.d]$ sudo cat example.conf 
> template: example
>       on: random
>   lookup: max -10s match-ids foreach random1,random2,random3
>     warn: $this > (($status >= $WARNING)  ? (75) : (85))
>     crit: $this > (($status == $CRITICAL) ? (85) : (95))
>       to: sysadmin
> ```
> 
> > lookup: max -10s match-ids foreach random1,random2,random3
> 
> I expect to to get 3 alarms with following `lookup dimensions`:
> 
>     * "lookup_dimensions": "random1"
> 
>     * "lookup_dimensions": "random2"
> 
>     * "lookup_dimensions": "random3"
> 
> 
> I got
> 
> ```json
>     "alarms": {
>         "example.random.example_random1": {
>             "id": 1565354639,
>             "name": "example_random1",
>             "chart": "example.random",
>             "lookup_dimensions": "random1",
>             "value": 293
>         },
>         "example.random.example_random2": {
>             "id": 1565354638,
>             "name": "example_random2",
>             "chart": "example.random",
>             "lookup_dimensions": "random1,random2",
>             "value": 293
>         },
>         "example.random.example_random3": {
>             "id": 1565354640,
>             "name": "example_random3",
>             "chart": "example.random",
>             "lookup_dimensions": "random1,random2,random3",
>             "value": 293
>         }
>     }
> ```

I changed the alarms used in the test to reproduce the errors found here and I fix all them, I also found other bug inside the code that was already fixed. Thanks!

thiagoftsm commented:
> i didnt mention it - the value is wrong too
> 
> ```json
>         "example.random.example_random1": {
>             "id": 1565354639,
>             "name": "example_random1",
>             "chart": "example.random",
>             "lookup_dimensions": "random1",
>         },
> ```
> 
> > "lookup_dimensions": "random1",
> 
> The value is sum of all dimensions.

There was a part of the code that I copied and paste without to change, this helped to create this error, but it was not the unique change necessary to fix this. Everything was already fixed.

thiagoftsm commented:
> ok, i tested the PR.
> 
> What i did:
> 
>     * installed the PR
> 
> 
> ```
> [ilyam@pc netdata_foreach]$ git status 
> On branch health_connection
> Your branch is up to date with 'origin/health_connection'.
> 
> nothing to commit, working tree clean
> ```
> 
>     * configured following alarm
> 
> 
> ```
> [ilyam@pc health.d]$ sudo cat example.conf 
> template: example
>       on: random
>   lookup: max -10s foreach *random*
>     warn: $this > (($status >= $WARNING)  ? (75) : (85))
>     crit: $this > (($status == $CRITICAL) ? (85) : (95))
>       to: sysadmin
> ```
> 
> Dimensions:
> 
> ```json
> {
>     "random1": {
>         "name": "random1"
>     },
>     "random2": {
>         "name": "random2"
>     },
>     "random3": {
>         "name": "random3"
>     }
> }
> ```
> 
> I expect to to get 3 alarms with following `lookup dimensions`:
> 
>     * "lookup_dimensions": "random1"
> 
>     * "lookup_dimensions": "random2"
> 
>     * "lookup_dimensions": "random3"
> 
> 
> I got 1 alarm
> 
> ```
> {
>     "example.random.example_*random*": {
>         "id": 1565354637,
>         "name": "example_*random*",
>         "chart": "example.random",
>         "family": "random",
>         "active": true,
>         "lookup_dimensions": "*random*",
>     }
> }
> ```
> 
> Patterns after `foreach` are simple patterns.

When you gave an answer in the initial issue #2107, you copied and changed the user dimensions without to use and "simple pattern", so when I created this PR I had in my mind that patterns would not be used to create the alarms. I confess I saw the patterns in the eval.c file, but I never worked with them, so my suggestions is we move in front with this like it was in the example, and after I study the patterns I release an update for this feature. What is your opinion about my suggestion?

ilyam8 commented:
>  and after I study the patterns I release an update for this feature. What is your opinion about my suggestion?

:point_right: Simplepatterns in foreach are must have.
:point_right: The featue just doesnt make sense w/o it.

That is the gist - simplify/automate creating alarms for dimension **if dimension name is not known**.

If i know dimension name i can copy template myself.  

How the feature should work (from my point of view) - foreach is a speacial alarm/template, its job is to clone itself len(foreach) times (that is it, it is not alarm actually, it produces alarms!), it does it every **update_every** (not only once at start).

Lets take same template as i used to test the PR

```
template: example
      on: random
  lookup: max -10s foreach *random*
    warn: $this > (($status >= $WARNING)  ? (75) : (85))
    crit: $this > (($status == $CRITICAL) ? (85) : (95))
      to: sysadmin
```

 - find the chart with context `random`

for that chart:
 - check every dimension against pattern(s)
 - if match (lets assume we match `random1`):
   - check if `template: example_random1` is exist, if yes, continue, if no create

```
template: example_random1
      on: random
  lookup: max -10s foreach of random1
    warn: $this > (($status >= $WARNING)  ? (75) : (85))
    crit: $this > (($status == $CRITICAL) ? (85) : (95))
      to: sysadmin
```

Do this work every **update_every** - it is pretty common when dimensions are added in runtime.

thiagoftsm commented:
> > and after I study the patterns I release an update for this feature. What is your opinion about my suggestion?
> 
> point_right Simplepatterns in foreach are must have.
> point_right The featue just doesnt make sense w/o it.
> 
> That is the gist - simplify/automate creating alarms for dimension **if dimension name is not known**.
> 
> If i know dimension name i can copy template myself.
> 
> How the feature should work (from my point of view) - foreach is a speacial alarm/template, its job is to clone itself len(foreach) times (that is it, it is not alarm actually, it produces alarms!), it does it every **every** (not only once at start).
> 
> Lets take same template as i used to test the PR
> 
> ```
> template: example
>       on: random
>   lookup: max -10s foreach *random*
>     warn: $this > (($status >= $WARNING)  ? (75) : (85))
>     crit: $this > (($status == $CRITICAL) ? (85) : (95))
>       to: sysadmin
> ```
> 
> Workflow:
> 
>     * find a chart(s) with context `random`
> 
> 
> for that chart(s):
> 
>     * check every dimension against pattern(s)
> 
>     * if match (lets assume we match `random1`):
>       
>       * check if `template: example_random1` is exist, if yes, continue, if no create
> 
> 
> ```
> template: example_random1
>       on: random
>   lookup: max -10s foreach of random1
>     warn: $this > (($status >= $WARNING)  ? (75) : (85))
>     crit: $this > (($status == $CRITICAL) ? (85) : (95))
>       to: sysadmin
> ```
> 
> Do this work every **every** - it is pretty common when dimensions are added in runtime.

I understood your point of view, I will study during the weekend how to query the dimensions of a chart in our database to return in this PR to bring this too.

Thanks by your answer!

thiagoftsm commented:
Hi @ilyam8 ,

Considering your last comments I executed a deep study in the libnetdata/simple_pattern/simple_pattern.c and in the database files, right now we have the following situation when we are creating alarms:

- Netdata starts from the main function and it will call `rrd_init`, this function calls `rrdhost_create` to create the localhost structure, inside this last function we have a call for `health_readdir` that  will read all the files with alarms and templates. At this point we do not have children threads started, we only have the main thread.
- When the children threads are created, we will have few threads creating charts(RRDSET),  and it case the conditions are matched inside some threads the function `rrdcalctemplate_link_matching` create other charts using the template. These two things happen in parallel

When I executed the first commits,  I did not have enough knowledge of the database process to create charts and dimension, this was one of the motives I could not create a function that would create alarms based in `simple patterns`. I am changing the direction now. The next  commits will bring the creation of the new alarms at the end of the process, when the dimensions are created, before to create the specific alarms, I will check whether the alarms associated to the chart is listed in the `foreach`explicitly or with `simple patterns`,  case it is and it was not created yet, I will create the new alarm.

ilyam8 commented:
> The next commits will bring the creation of the new alarms at the end of the process, when the dimensions are created

charts, dims can be created at any time, so you need to re do it based on some interval or event

thiagoftsm commented:
> > The next commits will bring the creation of the new alarms at the end of the process, when the dimensions are created
> 
> charts, dims can be created at any time, so you need to re do it based on some interval or event

For this specific PR, I think it is more necessary to pay attention for dimensions, because we are creating alarms depending on the existence of the dimension, but I will check charts too, to be sure we are not missing nothing.

thiagoftsm commented:
Hi @ilyam8 ,

I finished this PR doing functional tests with it right now. I am bringing 6 alarms to test the foreach, and a couple of alarms is using simple pattern to create charts for all dimensions.

ilyam8 commented:
@thiagoftsm is this ready to be tested?

thiagoftsm commented:
> @thiagoftsm is this ready to be tested?

Yes, it is already ready, please, test when you have a time.

thiagoftsm commented:
The last commit is only to update the branch with the latest merges that we execute, I did not change the code of the PR.

ilyam8 commented:
I started to test, it seems something is wrong.

my alarm

```
template: example
      on: random
  lookup: max -5s foreach *random*
  #lookup: max -5s of random1
  every:  5s
    warn: $this > (($status >= $WARNING)  ? (175) : (185))
    crit: $this > (($status == $CRITICAL) ? (185) : (195))
      to: sysadmin
```

Result (see `alarm not found -`):

<img width="948" alt="Screenshot 2019-09-17 at 16 33 53" src="https://user-images.githubusercontent.com/22274335/65046417-3fff4c00-d969-11e9-8bbc-083366e0c260.png">


Probably related

> 2019-09-17 09:36:52: netdata ERROR : MAIN : Cannot unlink alarm 'example.random.example' from host 'debian10': not found

thiagoftsm commented:
> > 2019-09-17 09:36:52: netdata ERROR : MAIN : Cannot unlink alarm 'example.random.example' from host 'debian10': not found

Hi @ilyam8 ,

thank you very much, there was a wrong calculation of hash inside the code, I fixed it and I did not see more both errors, please, can you verify that everything is correct now?

ilyam8 commented:
@thiagoftsm 	

seems to be working fine after the last change

what i did:
 - created the alarn

```
template: example
      on: random
  lookup: max -5s foreach random1 random2,!random4 *
    calc: $this/10
  every:  5s
    warn: $this > (($status >= $WARNING)  ? (175) : (185))
    crit: $this > (($status == $CRITICAL) ? (185) : (195))
      to: sysadmin
```

 - created 2 python jobs (example module)

> example.conf

```yaml
job1:
 var: 1

job2:
 var: 2
```

 - modified `example.chart.py`
   - add 2 dims after 15 iterations from start
   - add 2 more dims after 25 iterations from start
   - add 2 more dims after 35 iterations from start


<details>
<summary>python code</summary>

```python
class Service(SimpleService):
    def __init__(self, configuration=None, name=None):
        SimpleService.__init__(self, configuration=configuration, name=name)
        self.order = ORDER
        self.definitions = CHARTS
        self.random = SystemRandom()
        self.v = 0
        self.n = 4

    @staticmethod
    def check():
        return True


    def get_data(self):
        data = dict()
        self.v += 1
        if self.v == 15:
            self.n = 6
        if self.v == 25:
            self.n = 8
        if self.v == 35:
            self.n = 10

        for i in range(1, self.n):
            dimension_id = ''.join(['random', str(i)])

            if dimension_id not in self.charts['random']:
                self.charts['random'].add_dimension([dimension_id])

            data[dimension_id] = self.random.randint(0, 100)

        return data
```
</details>


Result:
 - All alarms in `Alarms->All`
 - All alarms in `/api/v1/alarms?all`

___

**Important**:
I tested only functionality. I didn't look at the code and there is no guarantee the is no regression/bugs.

@vlvkobal @mfundul check the code please

ilyam8 commented:
@thiagoftsm 	
and i still get

> 2019-09-18 07:10:40: netdata ERROR : MAIN : Cannot unlink alarm 'example_job2.random.example' from host 'debian10': not found

On every netdata restart.

thiagoftsm commented:
> > 2019-09-18 07:10:40: netdata ERROR : MAIN : Cannot unlink alarm 'example_job2.random.example' from host 'debian10': not found
> 
> On every netdata restart.

I fixed it now, when we have template scenario, the alarms are added to the index in a second stage, this is the motive we were seeing the messages, I added a new condition for we do not have more this false positive.

cakrit commented on health/README.md:
```diff
@@ -190,6 +190,11 @@ Everything is the same with [badges](../web/api/badges/). In short:
      have spaces in their names). This accepts Netdata simple patterns and the `match-ids` and
      `match-names` options affect the searches for dimensions.
 
+- `foreach DIMENSIONS` is optional and it will always be the last parameter. The rules for dimension here follow the same
```
These comments are very confusing. Please get help from @joelhans or @prhomhyse to improve the content here.

joelhans replied:
Here is my recommendation for altering the grammar/structure of this paragraph. This one sentence—`It is used to set the group of the dimensions of a chart that will have the same rule to raise an alarm.`—is difficult for me to understand and I may have rewritten it wrong. Let's make sure it's right before any changes are made.

> `foreach DIMENSIONS` is optional, will always be the last parameter, and uses the same `,`/`|` rules as the `of` parameter. Each dimension you specify in `foreach` will use the same rule to trigger an alarm. If you set both `of` and `foreach`, Netdata will ignore the `of` parameter and replace it with one of the dimensions you gave to `foreach`.

I'm having trouble understanding how I would try to format this line if I were going to use each of these parameters. @thiagoftsm maybe you could write an example template using `lookup` and `foreach` and we could document what each parameter accomplishes? We could add this to the **Examples** section, perhaps. It may also just help me better understand how to best document this rule.

thiagoftsm replied:
Hi @joelhans ,

I copied and paste your suggestion in the documentation, I also created two examples in the section cited, in the examples I decided to focus in the `lookup` line that is the greatest difference from the others, but case you find that is good to describe each line I can do.
Thank you to help me here!

ilyam8 commented on health/README.md:
```diff
@@ -665,6 +665,43 @@ Note that the drops chart does not exist if a network interface has never droppe
+    os: linux
+lookup: average -3s percentage foreach system,user
+ units: %
+ every: 1s
```
minimum recommended `every` is 10s, check other exanmples

thiagoftsm replied:
Thank you @ilyam8 , I fixed in both examples.

prhomhyse commented on health/README.md:
````diff
@@ -665,6 +665,43 @@ Note that the drops chart does not exist if a network interface has never droppe
+  crit: $this > 80
+```
+
+The `lookup` line will calculate the average of CPU usage from system and user in the last 3 seconds, due the fact we have
````
The `lookup` line will calculate the average of CPU usage from system and user in the last 3 seconds, due to the fact we have

thiagoftsm replied:
Thank you @prhomhyse , I overwrite the whole text with your suggestion.

prhomhyse commented on health/README.md:
````diff
@@ -675,12 +675,12 @@ Check if user or system dimension is using more than 50% of cpu:
 ```
 
-The `lookup` line will calculate the average of CPU usage from system and user in the last 3 seconds, due the fact we have
+The lookup line will calculate the average CPU usage from system and user in the last 3 seconds, due to the fact we have
````
**lookup** should be an inline code `lookup`:

Add backticks to it, it will look like the previous commit.

thiagoftsm replied:
When I pasted your correction I lost it, thank you for the review, I brought it back now.

joelhans commented on health/README.md:
````diff
@@ -660,6 +665,43 @@ Note that the drops chart does not exist if a network interface has never droppe
+```
+
+The `lookup` line will calculate the average of CPU usage from system and user in the last 3 seconds, here we will
+have the same behavior of the previous example, we will create alarms for all dimensions of the chart.
````
```suggestion
In this case, Netdata will create alarms for all dimensions of the chart.
```

thiagoftsm replied:
I fixed all the grammar according your suggestions.

thiagoftsm commented on health/README.md:
```diff
@@ -660,6 +665,43 @@ Note that the drops chart does not exist if a network interface has never droppe
+
+### Example 6
+
+Check if all dimensions is using more than 50% of cpu:
```
I copied and paste, and I forgot to change, thanks!

thiagoftsm commented on health/README.md:
````diff
@@ -660,6 +665,43 @@ Note that the drops chart does not exist if a network interface has never droppe
+```
+
+The `lookup` line will calculate the average CPU usage from system and user in the last 3 seconds, due to the fact we have
+the foreach in the `lookup` line, Netdata will create two independent alarms called respectively `dim_template_system`
````
Here I imported the Portuguese to English, thank you very much!

vlvkobal commented:
When I execute netdata with valgrind
```
sudo systemctl stop netdata; sudo rm -rf /var/log/netdata/*; sudo valgrind --leak-check=yes /usr/sbin/netdata -P /var/run/netdata/netdata.pid -D -W set global process scheduling policy keep -W set global OOM score keep
```
there is a huge amount of `Invalid read`/`Invalid free` errors on Netdata exit:
[valgrind.log](https://github.com/netdata/netdata/files/3641279/valgrind.log)

vlvkobal commented:
Is it correct to keep alarms up after dimensions have been obsoleted?
```
		{
			"id": "system.message_queue_messages",
			"name": "system.message_queue_messages",
			"type": "system",
			"family": "ipc message queues",
			"context": "system.message_queue_messages",
			"title": "IPC Message Queue Number of Messages (system.message_queue_messages)",
			"priority": 1201,
			"plugin": "proc.plugin",
			"module": "ipc",
			"enabled": true,
			"units": "messages",
			"data_url": "/api/v1/data?chart=system.message_queue_messages",
			"chart_type": "stacked",
			"duration": 541,
			"first_entry": 1569233684,
			"last_entry": 1569234224,
			"update_every": 1,
			"dimensions": {
				"5": { "name": "5" },
				"4": { "name": "4" },
				"0": { "name": "0" }
			},
			"chart_variables": {
			},
			"green": null,
			"red": null,
			"alarms": {
				"message_queues_0": {
					"id": 1569232196,
					"status": "CLEAR",
					"units": "messages",
					"update_every": 2
				},
				"message_queues_1": {
					"id": 1569232195,
					"status": "CLEAR",
					"units": "messages",
					"update_every": 2
				},
				"message_queues_2": {
					"id": 1569232194,
					"status": "CLEAR",
					"units": "messages",
					"update_every": 2
				},
				"message_queues_3": {
					"id": 1569232193,
					"status": "CLEAR",
					"units": "messages",
					"update_every": 2
				},
				"message_queues_4": {
					"id": 1569232192,
					"status": "CLEAR",
					"units": "messages",
					"update_every": 2
				},
				"message_queues_5": {
					"id": 1569232191,
					"status": "CLEAR",
					"units": "messages",
					"update_every": 2
				}
			}
		}
```
@ilyam8, what do you think?

ilyam8 commented:
I think this should be added in a separate PR.

thiagoftsm commented:
> I think this should be added in a separate PR.

I am working now in the valgrind error, and I will create another PR for this.

joelhans commented:
Thanks for working with me on the documentation, @thiagoftsm!

Docs LGTM at this point.

thiagoftsm commented:
Hi @vlvkobal ,

I fixed all the memory leak that was present in this PR, please verify again when you have a time.

Thank you for your help here.

thiagoftsm commented:
@vlvkobal now I fixed the problem with reloads, I only left open the problem that you discussed with @ilyam8 and I accepted his suggestion to open a new PR for it, all the other errors were removed.

cakrit commented:
Please clean up any addressed comments and provide your final review. If nothing else is pending, it should be merged tomorrow.

thiagoftsm commented:
> Please clean up any addressed comments and provide your final review. If nothing else is pending, it should be merged tomorrow.

All right, I am only waiting the final tests from others for I merge.

vlvkobal commented:
Valgrind errors on Netdata exit:
```
==164363== Memcheck, a memory error detector
==164363== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==164363== Using Valgrind-3.14.0 and LibVEX; rerun with -h for copyright info
==164363== Command: /usr/sbin/netdata -P /var/run/netdata/netdata.pid -D -W set global process scheduling policy keep -W set global OOM score keep
==164363==
^C==164363==
==164363== HEAP SUMMARY:
==164363==     in use at exit: 1,748,118 bytes in 38,699 blocks
==164363==   total heap usage: 269,497 allocs, 230,798 frees, 452,703,049 bytes allocated
==164363==
==164363== 1 bytes in 1 blocks are definitely lost in loss record 6 of 10,580
==164363==    at 0x483877F: malloc (vg_replace_malloc.c:299)
==164363==    by 0x58F936E: strdup (in /usr/lib/libc-2.29.so)
==164363==    by 0x153DD3: strdupz (libnetdata.c:166)
==164363==    by 0x161D62: poll_add_fd (socket.c:1233)
==164363==    by 0x161D62: poll_events_process (socket.c:1462)
==164363==    by 0x161D62: poll_events (socket.c:1687)
==164363==    by 0x1C34D2: socket_listen_main_static_threaded_worker (static-threaded.c:392)
==164363==    by 0x1C39C4: socket_listen_main_static_threaded (static-threaded.c:500)
==164363==    by 0x1650B9: thread_start (threads.c:153)
==164363==    by 0x585457E: start_thread (in /usr/lib/libpthread-2.29.so)
==164363==    by 0x596A0E2: clone (in /usr/lib/libc-2.29.so)
==164363==
==164363== 2 bytes in 2 blocks are definitely lost in loss record 446 of 10,580
==164363==    at 0x483877F: malloc (vg_replace_malloc.c:299)
==164363==    by 0x58F936E: strdup (in /usr/lib/libc-2.29.so)
==164363==    by 0x153DD3: strdupz (libnetdata.c:166)
==164363==    by 0x160FB7: poll_add_fd (socket.c:1233)
==164363==    by 0x160FB7: poll_events (socket.c:1606)
==164363==    by 0x1C34D2: socket_listen_main_static_threaded_worker (static-threaded.c:392)
==164363==    by 0x1C39C4: socket_listen_main_static_threaded (static-threaded.c:500)
==164363==    by 0x1650B9: thread_start (threads.c:153)
==164363==    by 0x585457E: start_thread (in /usr/lib/libpthread-2.29.so)
==164363==    by 0x596A0E2: clone (in /usr/lib/libc-2.29.so)
==164363==
==164363== 4 bytes in 4 blocks are definitely lost in loss record 846 of 10,580
==164363==    at 0x483877F: malloc (vg_replace_malloc.c:299)
==164363==    by 0x58F936E: strdup (in /usr/lib/libc-2.29.so)
==164363==    by 0x153DD3: strdupz (libnetdata.c:166)
==164363==    by 0x160FB7: poll_add_fd (socket.c:1233)
==164363==    by 0x160FB7: poll_events (socket.c:1606)
==164363==    by 0x1B2BA9: statsd_collector_thread (statsd.c:1011)
==164363==    by 0x1650B9: thread_start (threads.c:153)
==164363==    by 0x585457E: start_thread (in /usr/lib/libpthread-2.29.so)
==164363==    by 0x596A0E2: clone (in /usr/lib/libc-2.29.so)
==164363==
==164363== 5 bytes in 5 blocks are definitely lost in loss record 1,635 of 10,580
==164363==    at 0x483877F: malloc (vg_replace_malloc.c:299)
==164363==    by 0x58F936E: strdup (in /usr/lib/libc-2.29.so)
==164363==    by 0x153DD3: strdupz (libnetdata.c:166)
==164363==    by 0x161D62: poll_add_fd (socket.c:1233)
==164363==    by 0x161D62: poll_events_process (socket.c:1462)
==164363==    by 0x161D62: poll_events (socket.c:1687)
==164363==    by 0x1C34D2: socket_listen_main_static_threaded_worker (static-threaded.c:392)
==164363==    by 0x1650B9: thread_start (threads.c:153)
==164363==    by 0x585457E: start_thread (in /usr/lib/libpthread-2.29.so)
==164363==    by 0x596A0E2: clone (in /usr/lib/libc-2.29.so)
==164363==
==164363== 8 bytes in 1 blocks are definitely lost in loss record 2,667 of 10,580
==164363==    at 0x483877F: malloc (vg_replace_malloc.c:299)
==164363==    by 0x58F936E: strdup (in /usr/lib/libc-2.29.so)
==164363==    by 0x153DD3: strdupz (libnetdata.c:166)
==164363==    by 0x15B5C8: parse_pattern (simple_pattern.c:60)
==164363==    by 0x15B5C8: simple_pattern_create (simple_pattern.c:140)
==164363==    by 0x1835C3: health_pattern_from_foreach (health_config.c:330)
==164363==    by 0x195B49: rrdcalc_create_from_template (rrdcalc.c:388)
==164363==    by 0x1969A3: rrdcalctemplate_link_matching_test (rrdcalctemplate.c:17)
==164363==    by 0x196AAB: rrdcalctemplate_link_matching (rrdcalctemplate.c:36)
==164363==    by 0x19FF8C: rrdset_create_custom (rrdset.c:756)
==164363==    by 0x18CB34: pluginsd_process (plugins_d.c:456)
==164363==    by 0x18DB99: pluginsd_worker_thread (plugins_d.c:739)
==164363==    by 0x1650B9: thread_start (threads.c:153)
==164363==
==164363== 10 bytes in 10 blocks are definitely lost in loss record 3,126 of 10,580
==164363==    at 0x483877F: malloc (vg_replace_malloc.c:299)
==164363==    by 0x58F936E: strdup (in /usr/lib/libc-2.29.so)
==164363==    by 0x153DD3: strdupz (libnetdata.c:166)
==164363==    by 0x160FB7: poll_add_fd (socket.c:1233)
==164363==    by 0x160FB7: poll_events (socket.c:1606)
==164363==    by 0x1C34D2: socket_listen_main_static_threaded_worker (static-threaded.c:392)
==164363==    by 0x1650B9: thread_start (threads.c:153)
==164363==    by 0x585457E: start_thread (in /usr/lib/libpthread-2.29.so)
==164363==    by 0x596A0E2: clone (in /usr/lib/libc-2.29.so)
==164363==
==164363== 40 bytes in 1 blocks are definitely lost in loss record 6,281 of 10,580
==164363==    at 0x483AB65: calloc (vg_replace_malloc.c:752)
==164363==    by 0x153EC5: callocz (libnetdata.c:182)
==164363==    by 0x15B577: parse_pattern (simple_pattern.c:58)
==164363==    by 0x15B577: simple_pattern_create (simple_pattern.c:140)
==164363==    by 0x1835C3: health_pattern_from_foreach (health_config.c:330)
==164363==    by 0x195B49: rrdcalc_create_from_template (rrdcalc.c:388)
==164363==    by 0x1969A3: rrdcalctemplate_link_matching_test (rrdcalctemplate.c:17)
==164363==    by 0x196AAB: rrdcalctemplate_link_matching (rrdcalctemplate.c:36)
==164363==    by 0x19FF8C: rrdset_create_custom (rrdset.c:756)
==164363==    by 0x18CB34: pluginsd_process (plugins_d.c:456)
==164363==    by 0x18DB99: pluginsd_worker_thread (plugins_d.c:739)
==164363==    by 0x1650B9: thread_start (threads.c:153)
==164363==    by 0x585457E: start_thread (in /usr/lib/libpthread-2.29.so)
==164363==
==164363== LEAK SUMMARY:
==164363==    definitely lost: 70 bytes in 24 blocks
==164363==    indirectly lost: 0 bytes in 0 blocks
==164363==      possibly lost: 0 bytes in 0 blocks
==164363==    still reachable: 1,748,048 bytes in 38,675 blocks
==164363==         suppressed: 0 bytes in 0 blocks
==164363== Reachable blocks (those to which a pointer was found) are not shown.
==164363== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==164363==
==164363== For counts of detected and suppressed errors, rerun with: -v
==164363== ERROR SUMMARY: 7 errors from 7 contexts (suppressed: 0 from 0)
```

thiagoftsm commented:
I remember something now, the errors that you gave me I had them before the last commit, did you get the latest commit? I am seeing Codacy messages now.

vlvkobal commented:
valgrind-3.14.0

vlvkobal commented:
```
commit 2839959f9ad90159e72119b43df22f07bb711833 (HEAD -> health_connection, origin/health_connection)
Author: Thiago Marques <thiagoftsm@gmail.com>
Date:   Tue Sep 24 02:41:49 2019 +0000

    health_connection: Reload

    This commit fix the problem that the alarms were not linked after
    to receive a SIGUSR2
```

thiagoftsm commented:
I commit a new version now to avoid the false positive of codacy, the variables that it says not to be '\0' terminated are NULL terminated, the buffers are allocated using `callocz`.
I saw that you are one version behind me, I will run valgrind at least half hour to confirm that I can reproduce the same error you are receiving.

vlvkobal commented:
You don't need to run valgrind for a long time. It's enough to wait until alarms created from a template have started and terminate the session.

vlvkobal commented:
I've run more tests with valgrind. There are no errors only if alarms from the `foreach` template are not active.

thiagoftsm commented:
> I've run more tests with valgrind. There are no errors only if alarms from the `foreach` template are not active.

Thanks, I will check this now.

vlvkobal commented on database/rrdcalc.c:
```diff
@@ -407,11 +552,13 @@ void rrdcalc_free(RRDCALC *rc) {
     freez(rc->source);
     freez(rc->units);
     freez(rc->info);
+    freez(rc->spdim);
```
Sould be `simple_pattern_free(rc->spdim);`

vlvkobal replied:
No memory leaks after applying it.

thiagoftsm replied:
It is fixed now.

vlvkobal commented on health/health_json.c:
```diff
@@ -184,6 +184,9 @@ static inline void health_rrdcalc2json_nolock(RRDHOST *host, BUFFER *wb, RRDCALC
             health_string2json(wb, "\t\t\t", "lookup_dimensions", rc->dimensions, ",\n");
 
+        if(rc->foreachdim && *rc->foreachdim)
+            health_string2json(wb, "\t\t\t", "foreach", rc->foreachdim, ",\n");
```
I don't understand why, but I don't get this string in the response:
```
{
	"hostname": "arch-esxi",
	"latest_alarm_log_unique_id": 1560020145,
	"status": true,
	"now": 1569409900,
	"alarms": {
		"example.random.example_random3": {
			"id": 1569402711,
			"name": "example_random3",
			"chart": "example.random",
			"family": "random",
			"active": true,
			"disabled": false,
			"silenced": false,
			"exec": "/usr/libexec/netdata/plugins.d/alarm-notify.sh",
			"recipient": "sysadmin",
			"source": "2@/etc/netdata/health.d/example.conf",
			"units": "random number",
			"info": "",
			"status": "WARNING",
			"last_status_change": 1569409900,
			"last_updated": 1569409900,
			"next_update": 1569409902,
			"update_every": 2,
			"delay_up_duration": 0,
			"delay_down_duration": 0,
			"delay_max_duration": 0,
			"delay_multiplier": 1.000000,
			"delay": 0,
			"delay_up_to_timestamp": 1569409900,
			"warn_repeat_every": "0",
			"crit_repeat_every": "0",
			"value_string": "91 random number",
			"last_repeat": "0",
			"lookup_dimensions":"random3",
			"db_after": 1569409899,
			"db_before": 1569409899,
			"lookup_method": "average",
			"lookup_after": -1,
			"lookup_before": 0,
			"lookup_options": "",
			"warn":"$this &gt; (($status &gt;= $WARNING) ? (50) : (80))",
			"warn_parsed":"(${this} &gt; ((${status} &gt;= ${WARNING}) ? 50 : 80))",
			"crit":"$this &gt; (($status == $CRITICAL) ? (80) : (95))",
			"crit_parsed":"(${this} &gt; ((${status} == ${CRITICAL}) ? 80 : 95))",
			"green":null,
			"red":null,
			"value":91
		}
	}
}
```

thiagoftsm replied:
A very good find, thank you very much.
This is a dead code, when I started to develop I was keeping the `foreachdim` when I created the alarm, but after to exchange some messages with Ilya(Thank you Ilya), I was convinced that the `foreach` is a kind of template, and this specific variable is not more filled for normal alarms, they continue there only with their "templates".
I removed this code with the last commit.

thiagoftsm commented:
I ran Netdata without any alarm with `foreach` and I cannot see any new problem, please, can you share your log? Here it is my log

```
bash-5.0$ valgrind --leak-check=yes ./netdata -P /var/run/netdata/netdata.pid -D -W set global process scheduling policy keep -W set global OOM score keep                      
==3832== Memcheck, a memory error detector
==3832== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==3832== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
==3832== Command: ./netdata -P /var/run/netdata/netdata.pid -D -W set global process scheduling policy keep -W set global OOM score keep
==3832== 
^C^C==3832== Thread 3:
==3832== Syscall param pwrite64(buf) points to uninitialised byte(s)
==3832==    at 0x4AF2ED7: pwrite (in /lib64/libpthread-2.30.so)
==3832==    by 0x4ABB075: ??? (in /usr/lib64/libuv.so.1.0.0)
==3832==    by 0x4AB409D: ??? (in /usr/lib64/libuv.so.1.0.0)
==3832==    by 0x4AE806A: start_thread (in /lib64/libpthread-2.30.so)
==3832==    by 0x586172E: clone (in /lib64/libc-2.30.so)
==3832==  Address 0x7658961 is 2,401 bytes inside a block of size 4,096 alloc'd
==3832==    at 0x4839EB8: memalign (vg_replace_malloc.c:908)
==3832==    by 0x4839FCE: posix_memalign (vg_replace_malloc.c:1072)
==3832==    by 0x46AB56: wal_get_transaction_buffer (journalfile.c:83)
==3832==    by 0x4667A7: commit_data_extent (rrdengine.c:220)
==3832==    by 0x4668FE: do_commit_transaction (rrdengine.c:244)
==3832==    by 0x46776C: do_flush_pages (rrdengine.c:458)
==3832==    by 0x4689E0: rrdeng_worker (rrdengine.c:791)
==3832==    by 0x4AE806A: start_thread (in /lib64/libpthread-2.30.so)
==3832==    by 0x586172E: clone (in /lib64/libc-2.30.so)
==3832== 
==3832== 
==3832== HEAP SUMMARY:
==3832==     in use at exit: 1,070,614 bytes in 26,402 blocks
==3832==   total heap usage: 499,606 allocs, 473,204 frees, 55,372,502 bytes allocated
==3832== 
==3832== Thread 1:
==3832== 4 bytes in 4 blocks are definitely lost in loss record 892 of 9,750
==3832==    at 0x483777F: malloc (vg_replace_malloc.c:309)
==3832==    by 0x57E578A: strdup (in /lib64/libc-2.30.so)
==3832==    by 0x501311: strdupz (libnetdata.c:166)
==3832==    by 0x50DA24: poll_add_fd (socket.c:1233)
==3832==    by 0x50F035: poll_events (socket.c:1606)
==3832==    by 0x47CE50: statsd_collector_thread (statsd.c:1011)
==3832==    by 0x512169: thread_start (threads.c:153)
==3832==    by 0x4AE806A: start_thread (in /lib64/libpthread-2.30.so)
==3832==    by 0x586172E: clone (in /lib64/libc-2.30.so)
==3832==                                                                                                                                                                                                                                                                       
==3832== 7 bytes in 7 blocks are definitely lost in loss record 2,023 of 9,750                                                                                                                                                                                                 
==3832==    at 0x483777F: malloc (vg_replace_malloc.c:309)                                                                                                                                                                                                                     
==3832==    by 0x57E578A: strdup (in /lib64/libc-2.30.so)                                                                                                                                                                                                                      
==3832==    by 0x501311: strdupz (libnetdata.c:166)                                                                                                                                                                                                                            
==3832==    by 0x50DA24: poll_add_fd (socket.c:1233)                                                                                                                                                                                                                           
==3832==    by 0x50F035: poll_events (socket.c:1606)                                                                                                                                                                                                                           
==3832==    by 0x4922BD: socket_listen_main_static_threaded_worker (static-threaded.c:392)                                                                                                                                                                                     
==3832==    by 0x49289B: socket_listen_main_static_threaded (static-threaded.c:500)                                                                                                                                                                                            
==3832==    by 0x512169: thread_start (threads.c:153)                                                                                                                                                                                                                          
==3832==    by 0x4AE806A: start_thread (in /lib64/libpthread-2.30.so)                                                                                                                                                                                                          
==3832==    by 0x586172E: clone (in /lib64/libc-2.30.so)                                                                                                                                                                                                                       
==3832==                                                                                                                                                                                                                                                                       
==3832== 21 bytes in 21 blocks are definitely lost in loss record 4,772 of 9,750                                                                                                                                                                                               
==3832==    at 0x483777F: malloc (vg_replace_malloc.c:309)                                                                                                                                                                                                                     
==3832==    by 0x57E578A: strdup (in /lib64/libc-2.30.so)                                                                                                                                                                                                                      
==3832==    by 0x501311: strdupz (libnetdata.c:166)                                                                                                                                                                                                                            
==3832==    by 0x50DA24: poll_add_fd (socket.c:1233)                                                                                                                                                                                                                           
==3832==    by 0x50F035: poll_events (socket.c:1606)                                                                                                                                                                                                                           
==3832==    by 0x4922BD: socket_listen_main_static_threaded_worker (static-threaded.c:392)                                                                                                                                                                                     
==3832==    by 0x512169: thread_start (threads.c:153)                                                                                                                                                                                                                          
==3832==    by 0x4AE806A: start_thread (in /lib64/libpthread-2.30.so)                                                                                                                                                                                                          
==3832==    by 0x586172E: clone (in /lib64/libc-2.30.so)                                                                                                                                                                                                                       
==3832==                                                                                                                                                                                                                                                                       
==3832== LEAK SUMMARY:                                                                                                                                                                                                                                                         
==3832==    definitely lost: 32 bytes in 32 blocks                                                                                                                                                                                                                             
==3832==    indirectly lost: 0 bytes in 0 blocks                                                                                                                                                                                                                               
==3832==      possibly lost: 0 bytes in 0 blocks                                                                                                                                                                                                                               
==3832==    still reachable: 1,070,582 bytes in 26,370 blocks                                                                                                                                                                                                                  
==3832==         suppressed: 0 bytes in 0 blocks                                                                                                                                                                                                                               
==3832== Reachable blocks (those to which a pointer was found) are not shown.
==3832== To see them, rerun with: --leak-check=full --show-leak-kinds=all

```

vlvkobal commented:
Please read https://github.com/netdata/netdata/pull/6560#discussion_r328059049. There are no memory leaks if you use `simple_pattern_free` instead of `freez`.

thiagoftsm commented:
> Please read [#6560 (comment)](https://github.com/netdata/netdata/pull/6560#discussion_r328059049). There are no memory leaks if you use `simple_pattern_free` instead of `freez`.

Firstly I apologize, I used the correct code inside `rrdcalctemplate.c` and I did not change there before, so probably I missed the comments. I will pay more attention next time.
I hope with this last commit I fixed all memory leaks.

Thank you very much for your help here!
And for sure thank you llya!

thiagoftsm commented on tests/template_dimension/template_dim.sh.in:
```diff
@@ -0,0 +1,88 @@
+rm "$HEALTHDIR/dim_foreach_with_template_sp.conf"
+rm -rf "$OUTDIR"
+
+echo  -e "${GREEN} all the tests were sucessful"
```
I agree, it will be better return for the original color, I changed now.
